lib/puppet/environments.rb

```diff
@@ -350,7 +353,19 @@ def initialize(loader)
 
     # @!macro loader_list
     def list
-      @loader.list
+      # Evict all that have expired, in the same way as `get`
+      clear_all_expired
+
+      @loader.list.map do |env|
+        name = env.name
+        old_entry = @cache[name]
```

Comment:
there's nothing here we need to worry about regarding symbols vs strings?

Reply:
I walked through it and I guess strings vs symbols just worries me in general. But I don't see where it would have an effect.

Reply:
I kept the existing behavior of using whatever name is passed to `get`. Here we check if the [`name` is cached](https://github.com/puppetlabs/puppet/blob/72955cca3a8f8a99417c45ebef19c25ea2744376/lib/puppet/environments.rb#L367) and if not, we pass the [`name` to `add_entry`](https://github.com/puppetlabs/puppet/blob/72955cca3a8f8a99417c45ebef19c25ea2744376/lib/puppet/environments.rb#L376) and it [stores it under that key](https://github.com/puppetlabs/puppet/blob/72955cca3a8f8a99417c45ebef19c25ea2744376/lib/puppet/environments.rb#L384). Similarly, if we're asked to `clear` a specific entry, we just [use that `name`](https://github.com/puppetlabs/puppet/blob/72955cca3a8f8a99417c45ebef19c25ea2744376/lib/puppet/environments.rb#L402).

I agree the string vs symbol thing could bite us, so I filed https://tickets.puppetlabs.com/browse/PUP-10955